src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -318,46 +332,56 @@ func isCsrApproved(csr *certificatesv1.CertificateSigningRequest) bool {
 	return false
 }
 
-func (c controller) PostInstallConfigs(wg *sync.WaitGroup, status *ControllerStatus) {
-	defer wg.Done()
-	for {
-		time.Sleep(GeneralWaitInterval)
-		ctx := utils.GenerateRequestContext()
+func (c controller) PostInstallConfigs(ctx context.Context, wg *sync.WaitGroup, status *ControllerStatus) {
+	defer func() {
+		c.log.Infof("Finished PostInstallConfigs")
+		wg.Done()
+	}()
+	err := utils.WaitForPredicateWithContext(ctx, LongWaitTimeout, GeneralWaitInterval, func() bool {
+		ctxReq := utils.GenerateRequestContext()
 		cluster, err := c.ic.GetCluster(ctx)
```

Comment:
There are 3 different goroutine here that query the service independently. I would have this query in a single goroutine and communicate the result to other threads via a channel

Reply:
optimazation. can be done another day

Reply:
It's much more than mere optimization. It's quite a load. we will profile it and see the impact

Reply:
one api in a minute it is load? if yes we need to optimize service and not controller

Reply:
Agree we can add more optimizations but i don't think it is scope of this PR